homeassistant/core.py

```diff
@@ -814,6 +815,12 @@ def __init__(self, hass: HomeAssistant) -> None:
         """Initialize a new event bus."""
         self._listeners: dict[str, list[_FilterableJob]] = {}
         self._hass = hass
+        self._context_origin: WeakValueDictionary[str, Event] = WeakValueDictionary()
```

Comment:
The lru dict we use the recorder might be a better fit here.  I'm not 100 sure they will stay around long enough with a weakref

Reply:
The LRU definitely performs way better on heavy load.  Its only slightly heavier than the _LOGGER.debug is enabled check

Reply:
The LRU was faster but it got evicted too soon.